src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+            cv::Point2f imageCenter = std::accumulate(mCornersCache[i].begin(), mCornersCache[i].end(), cv::Point2f{}) / 4.0f;
+            mImmediateFiducials[id].id = id;
+            mImmediateFiducials[id].imageCenter = imageCenter;
+        }
+
+        // Remove fiducials in the immediate buffer that are no longer in sight
+        auto it = mImmediateFiducials.begin();
+        while (it != mImmediateFiducials.end()) {
+            if (std::find(mIdsCache.begin(), mIdsCache.end(), it->first) == mIdsCache.end()) {
+                it = mImmediateFiducials.erase(it);
+            } else {
+                ++it;
+            }
+        }
+
+        fiducial_msgs::FiducialTransformArray fidArray{};
+        fidArray.header.frame_id = ODOM_FRAME;
+        fidArray.header.stamp = ros::Time::now();
+        fidArray.header.seq = mSeqNum;
+        fidArray.transforms.reserve(mPersistentFiducials.size());
+
+        // Add readings to the persistent representations of the fiducials
+        for (auto [id, immediateFid]: mImmediateFiducials) {
+            PersistentFiducial& fid = mPersistentFiducials[id];
+            if (!immediateFid.fidInCam.has_value()) continue; // This is set if the point cloud had a valid reading for this fiducial
+
+            fid.id = id;
+            SE3 fidInOdom = SE3::transform(mTfBuffer, ROVER_FRAME, ODOM_FRAME, immediateFid.fidInCam.value(), mSeqNum);
+            fid.setFilterParams(mFilterCount, mFilterProportion);
+            fid.addReading(fidInOdom);
+        }
+
+        // Send all transforms of persistent fiducials
+        for (auto [id, fid]: mPersistentFiducials) {
+            if (!fid.fidInOdomX.ready()) continue; // Wait until the filters have enough readings to become meaningful
+
+            SE3::sendTransform(mTfBroadcaster, "fiducial" + std::to_string(id), ODOM_FRAME, fid.getPose(), mSeqNum);
+        }
+
+        mFidPub.publish(fidArray);
+
+        size_t detectedCount = mIdsCache.size();
+        if (mIsVerbose || !mPrevDetectedCount.has_value() || detectedCount != mPrevDetectedCount.value()) {
+            mPrevDetectedCount = detectedCount;
+            ROS_INFO("Detected %zu markers", detectedCount);
+        }
+
+        if (!mImmediateFiducials.empty()) {
+            cv::aruco::drawDetectedMarkers(mCvPtr->image, mCornersCache, mIdsCache);
+        }
+
+        if (mPublishImages) {
+            mImgPub.publish(mCvPtr->toImageMsg());
+        }
+
+        mSeqNum++;
+    } catch (cv_bridge::Exception const& e) {
+        ROS_ERROR("cv_bridge exception: %s", e.what());
+    } catch (cv::Exception const& e) {
+        ROS_ERROR("cv exception: %s", e.what());
+    }
+}
+
+/**
+ * @brief       Retrieve the pose of the fiducial in camera space
+ * @param msg   3D Point Cloud with points stored relative to the camera
+ * @param u     X Pixel Position
+ * @param v     Y Pixel Position
+ */
+SE3 getFidInCamFromPixel(sensor_msgs::PointCloud2ConstPtr const& msg, size_t u, size_t v) {
```

Comment:
ZED populates some values as NaN or Inf if we use the same mode we've been using (which is most accurate) https://www.stereolabs.com/docs/depth-sensing/depth-settings/. You will need to make this handle this here or it will cause your filter to explode down the line. This might need to return optional.

Reply:
True, this was handled properly in the old code, @JasonLiu2002 can you fix this?

Reply:
https://github.com/umrover/mrover-workspace/blob/6c84261ef3c8c8070da204fb0bbed64b93f10aca/jetson/percep/artag_detector.cpp#L185